docs/proposals/playbook_consolidation.md

```diff
@@ -0,0 +1,183 @@
+* `- include: playbooks/init/initialize_host_groups.yml` has been added to the
+  top of `playbooks/init/main.yml`
+* All other related files for initialization have been moved to `playbooks/init`
+
+The `initialize_host_groups.yml` playbook is only one play with one task for
+importing variables for inventory group conversions.  This task could be further
+consolidated with the play in `evaluate_groups.yml`.
+
+The new standard initialization playbook would be
+`playbooks/init/main.yml`.
+
+
+```
+ 
+> $ tree openshift-ansible/playbooks/init
+.
+├── evaluate_groups.yml
+├── initialize_facts.yml
+├── initialize_host_groups.yml
+├── initialize_openshift_repos.yml
+├── initialize_openshift_version.yml
+├── main.yml
+├── roles -> ../../roles
+├── validate_hostnames.yml
+└── vars
+    └── cluster_hosts.yml
+```
+
+```yaml
+# openshift-ansible/playbooks/init/main.yml
+---
+- include: initialize_host_groups.yml
+
+- include: evaluate_groups.yml
+
+- include: initialize_facts.yml
+
+- include: validate_hostnames.yml
+
+- include: initialize_openshift_repos.yml
+
+- include: initialize_openshift_version.yml
+```
+
+### `byo` and `common` Playbook Consolidation
+Historically, the `byo` directory coexisted with other platform directories
+which contained playbooks that then called into `common` playbooks to perform
+common installation steps for all platforms.  Since the other platform
+directories have been removed this separation is no longer necessary.
+
+In the example below, the following changes have been made:
+* `playbooks/byo/openshift-master` renamed to
+  `playbooks/openshift-master`
+* `playbooks/common/openshift-master` renamed to
+  `playbooks/openshift-master/private`
+* Original `byo` entry point playbooks have been updated to include their
+  respective playbooks from `private/`.
+* Symbolic links have been updated as necessary
+
+All user consumable playbooks are in the root of `openshift-master` and no entry
+point playbooks exist in the `private` directory.  Maintaining the separation
+between entry point playbooks and the private playbooks allows individual pieces
+of the deployments to be used as needed by other components.
+
+```
+openshift-ansible/playbooks/openshift-master 
+> $ tree
+.
+├── config.yml
+├── private
+│   ├── additional_config.yml
+│   ├── config.yml
+│   ├── filter_plugins -> ../../../filter_plugins
+│   ├── library -> ../../../library
+│   ├── lookup_plugins -> ../../../lookup_plugins
+│   ├── restart_hosts.yml
+│   ├── restart_services.yml
+│   ├── restart.yml
+│   ├── roles -> ../../../roles
+│   ├── scaleup.yml
+│   └── validate_restart.yml
+├── restart.yml
+└── scaleup.yml
+```
+
+```yaml
+# openshift-ansible/playbooks/openshift-master/config.yml
+---
+- include: ../init/main.yml
+
+- include: private/config.yml
+```
+
+The following examples shows how multiple components would be combined to
+perform a complete install.
+
+```yaml
+# openshift-ansible/playbooks/openshift-cluster/config.yml
```

Comment:
I feel this should be at openshift-ansible/playbooks/deploy_cluster.yml

There's no reason to put entry-point or top-level plays far inside child directories.

My feeling is 'openshift-cluster' is not a thing, and we should do away with it entirely.

Reply:
If `openshift-cluster` does not make sense as a 'thing', I'm fine with those playbooks being refactored. However, I don't think everything in `openshift-cluster` should be dumped in `playbooks/`.

Reply:
Additionally, in the proposed pattern, all entry point playbooks would be _only one_ level up from playbooks.  I too agree that we should not have entry point playbooks far inside child directories.

Reply:
I agree that not everything directly inside openshift-cluster should be dumped inside of playbooks.

But, I would like to see openshift-cluster go away and replaced by a file named at playbooks/deploy_cluster.yml.  deploy_cluster.yml would call into the various other plays (basically everything that is in openshift-cluster/private/config.yml) 

```yaml
# openshift-ansible/playbooks/deploy_cluster.yml
- include init/main.yml
- include openshift-etcd/private/config.yml
- ...
- include service_catalog/private/config.yml
```

All other 'entry-points' we provide would live in openshift-ansible/playbooks and would more or less resemble deploy_cluster.yml, except they are calling different set of plays/roles.  These other entry points would be only things that concern a subset of roles and plays that are not concerned with only one group of plays, like 'upgrade-control-plane.yml' would probably be a good candidate as it makes changes to more than one host group.

I like your idea of having the main.yml call the private/config.yml for all the other components, this will allow users to call just playbooks/openshift-master/main.yml if they so desire.

Reply:
If the entire solution was about single component cluster deployment, `openshift-ansible/playbooks/deploy_cluster.yml` would do perfectly fine. The following:

* `openshift-ansible/playbooks/scalup_master.yml`
* `openshift-ansible/playbooks/scalup_node.yml`
* `openshift-ansible/playbooks/scalup_etcd.yml`
* `openshift-ansible/playbooks/upgrade_master.yml`
* `openshift-ansible/playbooks/upgrade_node.yml`
...

will not lead to a well maintained hierarchy. We should emphasize which component is in question, i.e.

* `openshift-ansible/playbooks/master/scalup.yml`
* `openshift-ansible/playbooks/master/install/config.yml`
* `openshift-ansible/playbooks/master/upgrade.yml`
* `openshift-ansible/playbooks/node/scalup.yml`
* `openshift-ansible/playbooks/node/install/config.yml`
* `openshift-ansible/playbooks/node/upgrade.yml`

so it still make sense to have one extra directory. Unless you want to introduce the concept of:
* `openshift-ansible/playbooks/deploy_cluster.yml`: deploy entire cluster, i.e. masters, nodes, etcd, ...
* `openshift-ansible/playbooks/upgrade_cluster.yml`: upgrade entire cluster, i.e. masters, nodes, etcd, ...
* `openshift-ansible/playbooks/<OP>_cluster.yml`: carry `<OP>` operation over entire cluster

Reply:
> so it still make sense to have one extra directory. Unless you want to introduce the concept of:

    openshift-ansible/playbooks/deploy_cluster.yml: deploy entire cluster, i.e. masters, nodes, etcd, ...
    openshift-ansible/playbooks/upgrade_cluster.yml: upgrade entire cluster, i.e. masters, nodes, etcd, ...
    openshift-ansible/playbooks/<OP>_cluster.yml: carry <OP> operation over entire cluster

This seems sensible to me.